master/buildbot/locks.py

```diff
@@ -96,17 +96,19 @@ def isAvailable(self, requester, access):
 
         if access.mode == 'counting':
             # Wants counting access
-            return num_excl == 0 and num_counting + len(ahead) < self.maxCount \
+            return num_excl <= 1 - access.count \
+                and num_counting + len(ahead) + access.count <= self.maxCount \
                 and all([w[1].mode == 'counting' for w in ahead])
         # else Wants exclusive access
-        return num_excl == 0 and num_counting == 0 and not ahead
+        return num_excl == 0 and num_counting == 0 \
+            and not (ahead or not access.count)
 
     def _addOwner(self, owner, access):
         self.owners.append((owner, access))
         if access.mode == 'counting':
-            self._claimed_counting += 1
+            self._claimed_counting += access.count
         else:
-            self._claimed_excl += 1
+            self._claimed_excl += access.count
```

Comment:
I think it makes sense to not allow any other lock count than 1 for exclusive access. During construction of lock access, we should verify that correct access count has been passed.